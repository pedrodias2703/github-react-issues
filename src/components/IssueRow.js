import React from 'react';
import { Table } from 'semantic-ui-react';

import { IssueModel } from '../models/index';

const IssueRow = ({ issue = {} }) => {
  const printLabels = (labels) => {
    if (!labels.length) {
      return 'No labels';
    }
    return <li>{labels.map((label) => label.name)}</li>;
  };

  const recordIssue = new IssueModel({ ...issue });
  return (
    <Table.Row key={recordIssue.number}>
      <Table.Cell>
        <a href={recordIssue.html_url} target="_blank" rel="noopener noreferrer">
          {recordIssue.number}
        </a>
      </Table.Cell>
      <Table.Cell>{recordIssue.title}</Table.Cell>
      <Table.Cell>{new Date(recordIssue.created_at).toLocaleString()}</Table.Cell>
      <Table.Cell>{new Date(recordIssue.updated_at).toLocaleString()}</Table.Cell>
      <Table.Cell>{printLabels(recordIssue.labels)}</Table.Cell>
      <Table.Cell>{recordIssue.state.toUpperCase()}</Table.Cell>
    </Table.Row>
  );
};

export default IssueRow;
