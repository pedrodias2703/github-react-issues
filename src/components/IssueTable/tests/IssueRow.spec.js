import React from 'react';
import { shallow } from 'enzyme';

import IssueRow from '../Row';

const data = {
  url: 'https://api.github.com/repos/facebook/react/issues/14476',
  repository_url: 'https://api.github.com/repos/facebook/react',
  labels_url: 'https://api.github.com/repos/facebook/react/issues/14476/labels{/name}',
  comments_url: 'https://api.github.com/repos/facebook/react/issues/14476/comments',
  events_url: 'https://api.github.com/repos/facebook/react/issues/14476/events',
  html_url: 'https://github.com/facebook/react/issues/14476',
  id: 393051225,
  node_id: 'MDU6SXNzdWUzOTMwNTEyMjU=',
  number: 14476,
  title: 'useCallback/useEffect support custom comparator',
  labels: [
    {
      id: 1109410193,
      node_id: 'MDU6TGFiZWwxMTA5NDEwMTkz',
      url: 'https://api.github.com/repos/facebook/react/labels/Component:%20Hooks',
      name: 'Component: Hooks',
      color: 'c2f27b',
      default: false
    },
    {
      id: 40929155,
      node_id: 'MDU6TGFiZWw0MDkyOTE1NQ==',
      url: 'https://api.github.com/repos/facebook/react/labels/Type:%20Question',
      name: 'Type: Question',
      color: 'cc317c',
      default: false
    }
  ],
  state: 'open',
  locked: false,
  assignee: null,
  assignees: [],
  milestone: null,
  comments: 2,
  created_at: '2018-12-20T13:40:35Z',
  updated_at: '2018-12-20T16:40:00Z',
  closed_at: null,
  author_association: 'CONTRIBUTOR',
  body:
    'Currently we can pass an array as second argument when using `useCallback` or `useEffect` like below:\r\n\r\n```js\r\nuseCallback(()=> {\r\n  doSth(a, b)\r\n}, [a, b]) // how to do deep equal if a is an object ?\r\n```\r\n\r\nThe problem is it only compare array items with `===`,  it there any way to compare complex object ? \r\n\r\nSupport custom comparator as third argument looks not bad:\r\n\r\n```js\r\nuseCallback(()=> {\r\n  doSth(a, b)\r\n  }, \r\n  [complexObject], \r\n  (item, previousItem)=> { //custom compare logic, return true || false here }\r\n)\r\n```\r\n'
};

describe('<IssueRow />', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<IssueRow issue={data} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ul')).toHaveLength(1);
    expect(wrapper.find('li')).toHaveLength(data.labels.length);
    expect(wrapper.find('a')).toHaveLength(1);
  });

  it('should show no issues found message if data is empty', () => {
    const wrapper = shallow(<IssueRow issue={{}} />);
    expect(wrapper.html()).toContain('Standard Issue');
  });

  it('should show no issues found message if data is null', () => {
    const wrapper = shallow(<IssueRow />);
    expect(wrapper.html()).toContain('Standard Issue');
  });

  it('should print no labels if no labels available', () => {
    const newIssue = {
      ...data,
      labels: []
    };
    const wrapper = shallow(<IssueRow issue={newIssue} />);
    expect(wrapper.html()).toContain('No labels');
  });
});
