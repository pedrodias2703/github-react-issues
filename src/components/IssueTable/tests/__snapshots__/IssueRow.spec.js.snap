// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IssueRow /> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <IssueRow
    issue={
      Object {
        "assignee": null,
        "assignees": Array [],
        "author_association": "CONTRIBUTOR",
        "body": "Currently we can pass an array as second argument when using \`useCallback\` or \`useEffect\` like below:

\`\`\`js
useCallback(()=> {
  doSth(a, b)
}, [a, b]) // how to do deep equal if a is an object ?
\`\`\`

The problem is it only compare array items with \`===\`,  it there any way to compare complex object ? 

Support custom comparator as third argument looks not bad:

\`\`\`js
useCallback(()=> {
  doSth(a, b)
  }, 
  [complexObject], 
  (item, previousItem)=> { //custom compare logic, return true || false here }
)
\`\`\`
",
        "closed_at": null,
        "comments": 2,
        "comments_url": "https://api.github.com/repos/facebook/react/issues/14476/comments",
        "created_at": "2018-12-20T13:40:35Z",
        "events_url": "https://api.github.com/repos/facebook/react/issues/14476/events",
        "html_url": "https://github.com/facebook/react/issues/14476",
        "id": 393051225,
        "labels": Array [
          Object {
            "color": "c2f27b",
            "default": false,
            "id": 1109410193,
            "name": "Component: Hooks",
            "node_id": "MDU6TGFiZWwxMTA5NDEwMTkz",
            "url": "https://api.github.com/repos/facebook/react/labels/Component:%20Hooks",
          },
          Object {
            "color": "cc317c",
            "default": false,
            "id": 40929155,
            "name": "Type: Question",
            "node_id": "MDU6TGFiZWw0MDkyOTE1NQ==",
            "url": "https://api.github.com/repos/facebook/react/labels/Type:%20Question",
          },
        ],
        "labels_url": "https://api.github.com/repos/facebook/react/issues/14476/labels{/name}",
        "locked": false,
        "milestone": null,
        "node_id": "MDU6SXNzdWUzOTMwNTEyMjU=",
        "number": 14476,
        "repository_url": "https://api.github.com/repos/facebook/react",
        "state": "open",
        "title": "useCallback/useEffect support custom comparator",
        "updated_at": "2018-12-20T16:40:00Z",
        "url": "https://api.github.com/repos/facebook/react/issues/14476",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "14476",
    "nodeType": "function",
    "props": Object {
      "as": "tr",
      "cellAs": "td",
      "children": Array [
        <TableCell
          as="td"
        >
          <a
            href="https://github.com/facebook/react/issues/14476"
            rel="noopener noreferrer"
            target="_blank"
          >
            14476
          </a>
        </TableCell>,
        <TableCell
          as="td"
        >
          useCallback/useEffect support custom comparator
        </TableCell>,
        <TableCell
          as="td"
        >
          12/20/2018, 11:40:35 AM
        </TableCell>,
        <TableCell
          as="td"
        >
          12/20/2018, 2:40:00 PM
        </TableCell>,
        <TableCell
          as="td"
        >
          <ul>
            <li>
              Component: Hooks
            </li>
            <li>
              Type: Question
            </li>
          </ul>
        </TableCell>,
        <TableCell
          as="td"
        >
          OPEN
        </TableCell>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "td",
          "children": <a
            href="https://github.com/facebook/react/issues/14476"
            rel="noopener noreferrer"
            target="_blank"
          >
            14476
          </a>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 14476,
            "href": "https://github.com/facebook/react/issues/14476",
            "rel": "noopener noreferrer",
            "target": "_blank",
          },
          "ref": null,
          "rendered": 14476,
          "type": "a",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "td",
          "children": "useCallback/useEffect support custom comparator",
        },
        "ref": null,
        "rendered": "useCallback/useEffect support custom comparator",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "td",
          "children": "12/20/2018, 11:40:35 AM",
        },
        "ref": null,
        "rendered": "12/20/2018, 11:40:35 AM",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "td",
          "children": "12/20/2018, 2:40:00 PM",
        },
        "ref": null,
        "rendered": "12/20/2018, 2:40:00 PM",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "td",
          "children": <ul>
            <li>
              Component: Hooks
            </li>
            <li>
              Type: Question
            </li>
          </ul>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <li>
                Component: Hooks
              </li>,
              <li>
                Type: Question
              </li>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": "Component: Hooks",
              },
              "ref": null,
              "rendered": "Component: Hooks",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": "Type: Question",
              },
              "ref": null,
              "rendered": "Type: Question",
              "type": "li",
            },
          ],
          "type": "ul",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "td",
          "children": "OPEN",
        },
        "ref": null,
        "rendered": "OPEN",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "14476",
      "nodeType": "function",
      "props": Object {
        "as": "tr",
        "cellAs": "td",
        "children": Array [
          <TableCell
            as="td"
          >
            <a
              href="https://github.com/facebook/react/issues/14476"
              rel="noopener noreferrer"
              target="_blank"
            >
              14476
            </a>
          </TableCell>,
          <TableCell
            as="td"
          >
            useCallback/useEffect support custom comparator
          </TableCell>,
          <TableCell
            as="td"
          >
            12/20/2018, 11:40:35 AM
          </TableCell>,
          <TableCell
            as="td"
          >
            12/20/2018, 2:40:00 PM
          </TableCell>,
          <TableCell
            as="td"
          >
            <ul>
              <li>
                Component: Hooks
              </li>
              <li>
                Type: Question
              </li>
            </ul>
          </TableCell>,
          <TableCell
            as="td"
          >
            OPEN
          </TableCell>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "td",
            "children": <a
              href="https://github.com/facebook/react/issues/14476"
              rel="noopener noreferrer"
              target="_blank"
            >
              14476
            </a>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 14476,
              "href": "https://github.com/facebook/react/issues/14476",
              "rel": "noopener noreferrer",
              "target": "_blank",
            },
            "ref": null,
            "rendered": 14476,
            "type": "a",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "td",
            "children": "useCallback/useEffect support custom comparator",
          },
          "ref": null,
          "rendered": "useCallback/useEffect support custom comparator",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "td",
            "children": "12/20/2018, 11:40:35 AM",
          },
          "ref": null,
          "rendered": "12/20/2018, 11:40:35 AM",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "td",
            "children": "12/20/2018, 2:40:00 PM",
          },
          "ref": null,
          "rendered": "12/20/2018, 2:40:00 PM",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "td",
            "children": <ul>
              <li>
                Component: Hooks
              </li>
              <li>
                Type: Question
              </li>
            </ul>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <li>
                  Component: Hooks
                </li>,
                <li>
                  Type: Question
                </li>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "host",
                "props": Object {
                  "children": "Component: Hooks",
                },
                "ref": null,
                "rendered": "Component: Hooks",
                "type": "li",
              },
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "host",
                "props": Object {
                  "children": "Type: Question",
                },
                "ref": null,
                "rendered": "Type: Question",
                "type": "li",
              },
            ],
            "type": "ul",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "td",
            "children": "OPEN",
          },
          "ref": null,
          "rendered": "OPEN",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
